File: capture.py
import pyshark
import socket
import psutil
import struct
from datetime import datetime
import os
import logging
import sys
import shutil  # Add this import
from typing import Tuple, Optional
from google.protobuf.json_format import MessageToJson

# Add the absolute path to the protos directory
current_dir = os.path.dirname(os.path.abspath(__file__))
protos_path = os.path.join(current_dir, "networking", "protos")
sys.path.append(protos_path)

from networking.protos import Lobby_pb2
from networking.protos import _PacketCommand_pb2

class PacketProtocol:
    S2C_LOBBY_CHARACTER_INFO_RES = 44
    S2C_ACCOUNT_CHARACTER_LIST_RES = 18
    
    @staticmethod
    def is_valid_type(proto_type: int) -> bool:
        return proto_type in [
            PacketProtocol.S2C_LOBBY_CHARACTER_INFO_RES,
            PacketProtocol.S2C_ACCOUNT_CHARACTER_LIST_RES
        ]

class PacketCapture:
    def __init__(self, interface: str = 'Ethernet', port_range: Tuple[int, int] = (20200, 20300)):
        self.interface = interface
        self.port_range = port_range
        self.packet_data = b""
        logging.basicConfig(level=logging.INFO)
        self.logger = logging.getLogger(__name__)
        self.data_dir = "data"
        os.makedirs(self.data_dir, exist_ok=True)
        self.MAX_BUFFER_SIZE = 1024 * 1024  # Increase to 1MB
        self.expected_packet_length = None
        self.expected_proto_type = None

    def get_local_ip(self) -> Optional[str]:
        for interface, addrs in psutil.net_if_addrs().items():
            if interface == self.interface:
                for addr in addrs:
                    if addr.family == socket.AF_INET:
                        return addr.address
        return None

    def validate_packet_header(self, length: int, proto_type: int, padding: int) -> bool:
        """Validate packet header values"""
        valid_packet_range = (100, 2 * 1024 * 1024)  # Between 100 bytes and 2MB
        return (
            valid_packet_range[0] <= length <= valid_packet_range[1] and
            PacketProtocol.is_valid_type(proto_type) and
            padding in [0, 256]  # Common padding values
        )

    def process_packet(self, data: bytes) -> None:
        if len(data) > 0:
            # Add incoming data to buffer
            self.packet_data += data
            current_size = len(self.packet_data)
            
            # Reset if buffer gets too large
            if current_size > self.MAX_BUFFER_SIZE:
                print(f"Buffer exceeded max size ({self.MAX_BUFFER_SIZE} bytes)")
                self.reset_state()
                return

            # Try to parse/validate header
            if self.expected_packet_length is None and current_size >= 8:
                try:
                    packet_length, proto_type, random_padding = struct.unpack('<IHH', self.packet_data[:8])
                    
                    if not self.validate_packet_header(packet_length, proto_type, random_padding):
                        print(f"Invalid header: Length={packet_length}, Type={proto_type}, Padding={random_padding}")
                        self.reset_state()
                        return

                    print(f"New packet header: Length={packet_length}, Type={proto_type}, Padding={random_padding}")
                    self.expected_packet_length = packet_length
                    self.expected_proto_type = proto_type
                except struct.error:
                    self.reset_state()
                    return

            # Process packet data
            if self.expected_packet_length and self.expected_proto_type:
                # Handle overflow by trimming
                if current_size > self.expected_packet_length:
                    print(f"Trimming overflow {current_size} -> {self.expected_packet_length}")
                    self.packet_data = self.packet_data[:self.expected_packet_length]
                    current_size = self.expected_packet_length

                # Complete packet
                if current_size == self.expected_packet_length:
                    if self.verify_packet():
                        self.save_packet_data()
                    self.reset_state()
                # Progress update
                elif current_size % 8192 == 0:
                    print(f"Accumulating: {current_size}/{self.expected_packet_length}")

    def verify_packet(self) -> bool:
        """Verify packet integrity before saving"""
        try:
            if len(self.packet_data) != self.expected_packet_length:
                return False
                
            # Try parsing the protobuf message without saving
            data = self.packet_data[8:]
            info = Lobby_pb2.SS2C_LOBBY_CHARACTER_INFO_RES()
            info.ParseFromString(data)
            return True
        except Exception as e:
            print(f"Packet verification failed: {e}")
            return False

    def reset_state(self) -> None:
        """Reset all packet processing state"""
        self.packet_data = b""
        self.expected_packet_length = None
        self.expected_proto_type = None

    def save_packet_data(self) -> None:
        try:
            if len(self.packet_data) != self.expected_packet_length:
                raise ValueError("Packet length mismatch")
                
            data = self.packet_data[8:]
            info = Lobby_pb2.SS2C_LOBBY_CHARACTER_INFO_RES()
            info.ParseFromString(data)
            json_data = MessageToJson(info)
            
            # Save both to data dir and root for compatibility
            timestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
            data_file = os.path.join(self.data_dir, f"{timestamp}.json")
            root_file = "packet_data.json"
            
            # Save to timestamped file
            with open(data_file, "w", encoding='utf-8') as f:
                f.write(json_data)
                
            # Save/update packet_data.json in root
            with open(root_file, "w", encoding='utf-8') as f:
                f.write(json_data)
            
            print(f"Successfully saved packet data to {data_file} and {root_file}")
            
        except Exception as e:
            print(f"Failed to save packet data: {str(e)}")
            raise

    def capture(self) -> None:
        local_ip = self.get_local_ip()
        if not local_ip:
            print(f"Could not find IP address for interface {self.interface}")
            return

        display_filter = (f'ip.dst == {local_ip} and '
                         f'tcp.srcport >= {self.port_range[0]} and '
                         f'tcp.srcport <= {self.port_range[1]}')
        
        capture = pyshark.LiveCapture(interface=self.interface, display_filter=display_filter)

        try:
            for packet in capture.sniff_continuously():
                if 'TCP' in packet and hasattr(packet.tcp, 'payload'):
                    self.process_packet(packet.tcp.payload.binary_value)
        except KeyboardInterrupt:
            print("Capture stopped by user")
        except Exception as e:
            print(f"Capture error: {e}")
        finally:
            try:
                capture.close()
            except:
                pass
            self.packet_data = b""

    def _process_packet_wrapper(self, packet):
        if 'TCP' in packet and hasattr(packet.tcp, 'payload'):
            self.process_packet(packet.tcp.payload.binary_value)

def main():
    capture = PacketCapture()
    capture.capture()

if __name__ == "__main__":
    main()

File: nodejs/capture.js
// Node.js version of capture.py using 'cap' and protobufjs
const Cap = require('cap').Cap;
const decoders = require('cap').decoders;
const PROTOCOL = require('cap').PROTOCOL;
const os = require('os');
const fs = require('fs-extra');
const path = require('path');
const protobuf = require('protobufjs');

class PacketProtocol {
  static S2C_LOBBY_CHARACTER_INFO_RES = 44;
  static S2C_ACCOUNT_CHARACTER_LIST_RES = 18;
  static isValidType(type) {
    return [
      PacketProtocol.S2C_LOBBY_CHARACTER_INFO_RES,
      PacketProtocol.S2C_ACCOUNT_CHARACTER_LIST_RES
    ].includes(type);
  }
}

class PacketCapture {
  constructor(interfaceName = 'Ethernet', portRange = [20200, 20300]) {
    this.interfaceName = interfaceName;
    this.portRange = portRange;
    this.packetData = Buffer.alloc(0);
    this.MAX_BUFFER_SIZE = 1024 * 1024;
    this.expectedPacketLength = null;
    this.expectedProtoType = null;
    this.dataDir = path.resolve(__dirname, '../data');
    fs.ensureDirSync(this.dataDir);
  }

  getLocalIP() {
    const ifaces = os.networkInterfaces();
    const addrs = ifaces[this.interfaceName] || [];
    for (const addr of addrs) {
      if (addr.family === 'IPv4' && !addr.internal) {
        return addr.address;
      }
    }
    return null;
  }

  validatePacketHeader(length, protoType, padding) {
    return (
      length >= 100 && length <= 2 * 1024 * 1024 &&
      PacketProtocol.isValidType(protoType) &&
      [0, 256].includes(padding)
    );
  }

  async verifyPacket() {
    if (this.packetData.length !== this.expectedPacketLength) return false;
    try {
      const body = this.packetData.slice(8);
      const root = await protobuf.load(
        path.resolve(__dirname, '../networking/protos/Lobby.proto')
      );
      const Msg = root.lookupType('SS2C_LOBBY_CHARACTER_INFO_RES');
      Msg.decode(body);
      return true;
    } catch (err) {
      console.error('Packet verification failed:', err);
      return false;
    }
  }

  async savePacketData() {
    const body = this.packetData.slice(8);
    const root = await protobuf.load(
      path.resolve(__dirname, '../networking/protos/Lobby.proto')
    );
    const Msg = root.lookupType('SS2C_LOBBY_CHARACTER_INFO_RES');
    const message = Msg.decode(body);
    const json = Msg.toObject(message, { enums: String, longs: String, defaults: true });

    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const dataFile = path.join(this.dataDir, `${timestamp}.json`);
    const rootFile = path.resolve(__dirname, '../packet_data.json');

    await fs.writeJson(dataFile, json, { spaces: 2 });
    await fs.writeJson(rootFile, json, { spaces: 2 });
    console.log(`Saved packet data to ${dataFile} and ${rootFile}`);
  }

  async processPacket(dataBuf) {
    this.packetData = Buffer.concat([this.packetData, dataBuf]);
    if (this.packetData.length > this.MAX_BUFFER_SIZE) {
      console.warn('Buffer exceeded max size');
      this.resetState();
      return;
    }

    if (this.expectedPacketLength === null && this.packetData.length >= 8) {
      const length = this.packetData.readUInt32LE(0);
      const protoType = this.packetData.readUInt16LE(4);
      const padding = this.packetData.readUInt16LE(6);

      if (!this.validatePacketHeader(length, protoType, padding)) {
        console.warn(`Invalid header: len=${length}, type=${protoType}, pad=${padding}`);
        this.resetState();
        return;
      }
      console.log(`New packet header: len=${length}, type=${protoType}, pad=${padding}`);
      this.expectedPacketLength = length;
      this.expectedProtoType = protoType;
    }

    if (this.expectedPacketLength && this.expectedProtoType) {
      if (this.packetData.length > this.expectedPacketLength) {
        this.packetData = this.packetData.slice(0, this.expectedPacketLength);
      }

      if (this.packetData.length === this.expectedPacketLength) {
        if (await this.verifyPacket()) {
          await this.savePacketData();
        }
        this.resetState();
      }
    }
  }

  resetState() {
    this.packetData = Buffer.alloc(0);
    this.expectedPacketLength = null;
    this.expectedProtoType = null;
  }

  capture() {
    const localIP = this.getLocalIP();
    if (!localIP) {
      console.error(`Unable to find IP for interface ${this.interfaceName}`);
      return;
    }

    const cap = new Cap();
    const filter = `ip dst ${localIP} and tcp src portrange ${this.portRange[0]}-${this.portRange[1]}`;
    const bufSize = 10 * 1024 * 1024;
    const buffer = Buffer.alloc(65535);
    const linkType = cap.open(this.interfaceName, filter, bufSize, buffer);

    cap.setMinBytes && cap.setMinBytes(0);

    cap.on('packet', async (nbytes, trunc) => {
      const ret = decoders.Ethernet(buffer);
      if (ret.info.type === PROTOCOL.ETHERNET.IPV4) {
        const ret2 = decoders.IPV4(buffer, ret.offset);
        if (ret2.info.protocol === PROTOCOL.IP.TCP) {
          const ret3 = decoders.TCP(buffer, ret2.offset);
          if (ret3.info.payloadLength > 0) {
            const payload = buffer.slice(ret3.offset, ret3.offset + ret3.info.payloadLength);
            await this.processPacket(payload);
          }
        }
      }
    });

    console.log('Listening on', this.interfaceName, 'with filter', filter);
  }
}

// Run capture
const pc = new PacketCapture();
pc.capture();